### DISCUSSION TITLE
[Docs] Translating imperative control flows to statecharts

### DISCUSSION THREAD

**Discussion by ChrisShank** [statelyai/xstate#2355] [Ideas]
I think one of the early challenges of learning about `xstate` is that the control flow is different than imperative programming languages offer. I was wondering it would be useful to add a *translation* of imperative control flows to declarative statechart control flows to the xstate documentation.

Here are some examples:

State transition
```ts
let isLoading = false;

isLoading = false; // Transition to a new state
```

```ts
loading: {
  on: { RESOLVED: 'done' },
},
done: {},
```

Conditional transitions (permutations of `if-else` statements)

```ts
if (someCondition) {
  // update state
} else if (someOtherCondition) {
   // update some other state
} else {
  // update some other state
}
```

```ts
someState: {
  on: {
    someEvent: [
      { cond: 'someCondition', target: 'nextState' },
      { cond: 'someOtherCondition', target: 'someNextState' },
      { target: 'someOtherState' },
    ]
  }
}
```

**Comment by hrafnkellpalsson** [comment]
I literally did this yesterday for a coworker and headed over to the Discord to talk about, and found this Discussion! It seemed to work to help solidify his understanding of XState.

This also made me think of the gif Sunil Pai (I think) made when hooks were introduced in React.
The one that showed how code moved around from React lifecycle functions to hooks, maybe there's a way to do something similar for imperative -> xstate?

**Reply by ChrisShank** [reply]
That's great to hear! Do you have any good examples that we could discuss?

**Reply by hrafnkellpalsson** [reply]
How a dynamic initial value of useState/useReducer could be modeled as an always transition on an initial 'idle' state with guards to immediately send you to the desired state!

**Comment by robertpenner** [comment]
Great idea and a decent start on implementing.

Possible next steps:
- Account for events in the imperative example. `RESOLVED` and `someEvent` only exist in the XState version.
- Translate nested states
- Translate parallel states

For reference, a statechart is based on an and/or tree:

https://twitter.com/robpenner/status/1401531164536545282

![image](https://user-images.githubusercontent.com/79827/124327762-64d0ad80-db56-11eb-8963-4886802d1773.png)

**Comment by ChrisShank** [comment]
One example to translate `while` loops  could be batching behavior (i.e send a total of 100 network requests in batches of 5). The machine would be a self transition with a guard.

**Comment by ChrisShank** [comment]
Another example is converging parallel tasks like what ‘Promise.all` is used for. https://xstate.js.org/docs/guides/final.html#parallel-states

**Comment by ChrisShank** [comment]
Fizzbuzz as a statechart! https://statecharts.dev/fizzbuzz-actions-internal-events.html

What a completely different perspective about how to reason about the problem

**Reply by ChrisShank** [reply]
Here is my own implementation that is closer to imperative logic

```ts
import { assign, createMachine, interpret } from 'xstate';
import { log } from 'xstate/lib/actions';

import { inspect } from '@xstate/inspect';
inspect({ iframe: false });

const fizzBuzzMachine = createMachine<{ count: number }>(
  {
    id: 'fizzbuzz',
    initial: 'digit',
    context: { count: 1 },
    on: {
      increment: { target: 'calculating', actions: 'incrementCount' },
    },
    states: {
      calculating: {
        always: [
          { cond: 'is done', target: 'done' },
          { cond: 'is fizz buzz', target: 'fizz buzz' },
          { cond: 'is fizz', target: 'fizz' },
          { cond: 'is buzz', target: 'buzz' },
          { target: 'digit' },
        ],
      },
      fizz: { entry: log('fizz') },
      buzz: { entry: log('buzz') },
      'fizz buzz': { entry: log('fizz buzz') },
      digit: { entry: log(({ count }) => count) },
      done: { type: 'final' },
    },
  },
  {
    actions: {
      incrementCount: assign({ count: ({ count }) => count + 1 }),
    },
    guards: {
      'is done': ({ count }) => count > 100,
      'is fizz buzz': ({ count }) => count % 15 === 0,
      'is fizz': ({ count }) => count % 3 === 0,
      'is buzz': ({ count }) => count % 5 === 0,
    },
  }
);

const service = interpret(fizzBuzzMachine, { devTools: true }).start();

while (!service.state.done) service.send('increment');
```

**Comment by ChrisShank** [comment]
Its also interesting to see side by side what it looks like to deal with imperative code that changes over time bs xstate. This can be defined as the [minimum editing distance](https://www.freecodecamp.org/news/how-to-optimize-for-change-software-development/#minimize-edit-distance). For example,

1. Start with a simple fetch
2. Catch error and log it
3. Add retry logic
4. Handle responding to user input to retry
5. …
